### ISSUE TITLE
test_cc failures on 7.0.1

### ISSUE THREAD

**Issue by jelmer** [heimdal/heimdal#224]
test_cc is failing for me on some configurations. The error message is in the log file is a fairly useless sqlite error:

test_cc: krb5_cc_resolve: Error opening scache file /tmp/krb5scc_1000: unable to open database file
FAIL test_cc (exit status: 1)

I haven't been able to dig down further than that yet. All other tests are fine.

**Comment by jelmer** [comment]
FWIW this still happens with 7.0.2

**Comment by vdukhovni** [comment]
When this fails, what is the state of that /tmp/krb5scc_1000 file? Is it a valid sqlite3 database? Is the schema in place? ... More information needed.

We've upgraded the sqlite3 bits to 3.15.1, and 3.15.2 is out with some bugfixes, though unlikely, it may be worth trying to replace the bundled sqlite3 code with the version from 3.15.2.

**Comment by nicowilliams** [comment]
It is concerning that a test used a default ccache file from /tmp though.  I can confirm that `test_cc` actually does that, though it doesn't fail for me.  @vdukhovni's question is relevant therefore: did you have a pre-existing /tmp/krb5scc_1000?  But regardless, we should fix `test_cc` to not use a default ccache.  How?  I guess we could `mkstemp()` a file and `putenv()` `KRB5CCNAME`.

**Comment by jelmer** [comment]
THat was a good point. The "clean" chroot I was doing this build in had a remaining krb5scc_1000 file in /tmp, which is why I had so much trouble reproducing it...

**Comment by nicowilliams** [comment]
So now it passes?

**Comment by jelmer** [comment]
Yes, test_cc passes if I make sure that file is removed before I run the tests.

I now get failures further in the testsuite, but those appear to be unrelated.

**Comment by nicowilliams** [comment]
(Writing to /tmp is hadrly a problem, provided it's always done with `mkstemp[o]()`!

**Comment by jelmer** [comment]
I was thinking it'd be easier to just set a custom TMPDIR during tests than to modify all code that uses predictable filenames in /tmp at the moment.

**Comment by nicowilliams** [comment]
@jelmer Yes, that we should do, I agree, it's a good idea, though it may be non-trivial because many tests are C-coded and run straight from make, so we'd have to add a bit of harness to make a tmp tmpdir, set it, run the test, check that the test left no garbage, and remove the tmp tmpdir.  But `test_cc` shouldn't be using default ccaches even so.

**Comment by vdukhovni** [comment]
> @jelmer I now get failures further in the testsuite, but those appear to be unrelated.

At this point, I am more curious what those *further* failures might be. Please elaborate...

**Comment by nicowilliams** [comment]
@jelmer If we don't we hear about those by Saturday night then we'll ship rc3 + the OpenBSD LIB_pidfile fix as 7.1 on Monday.  It's OK, anyways, as the whole idea is that we'll now do new releases more often.  So we'll have a 7.2 soon after anyways.

**Comment by jelmer** [comment]
Sorry, both Brian and I are very busy at the moment. You can see current Heimdal build failures for Debian at https://buildd.debian.org/status/package.php?p=heimdal (see https://release.debian.org/stretch/arch_qualify.html for a list of architectures that matter for the release).

The remaining issues seem to be check-iprop failing on ppc64el (https://buildd.debian.org/status/fetch.php?pkg=heimdal&arch=ppc64&ver=7.0.2%2Bdfsg-1&stamp=1481772679) and failure to link hdb/test_dbinfo on mips* (missing symbol hdb_dbinfo_get_label - https://buildd.debian.org/status/fetch.php?pkg=heimdal&arch=mips&ver=7.0.2%2Bdfsg-1&stamp=1481776819).

There are also various failures on non-release architectures, but those aren't really an issue blocking inclusion of Heimdal in stretch.

**Comment by nicowilliams** [comment]
The `check-iprop` failure is almost certainly a race condition because we can only sleep in between certain operations, and there's one where we don't look for evidence in the messages.log.  I recommend that Debian patch `tests/kdc/check-iprop.in`.

As to the MIPS failure, I don't understand it at all.  That symbol is very much exported from `libhdb`: it's in `lib/hdb/version-script.map`, and it's there when I `nm -D` a built `libhdb.so`.  That function has been exported for a long time.  I don't know what might be wrong on MIPS.

**Comment by nicowilliams** [comment]
We might like to get more output from make check on these systems.  What would be a convenient way to do that?  `make check V=1` does not produce verbose output :(

Also, re: `check-iprop`, a proper fix for it so that it never sleeps too little will take significant time and effort.  If this is a make-or-break issue for Debian we'll fix it and re-issue 7.1, otherwise we'd rather leave it for 7.1.1.

**Comment by jelmer** [comment]
On Tue, Dec 20, 2016 at 06:57:54AM -0800, Nico Williams wrote:
> We might like to get more output from make check on these systems.  What would be a convenient way to do that?  `make check V=1` does not produce verbose output :(
Yeah, that's long annoyed me as well. :-( Having a way of automatically catting the test logs whne a test fails (https://github.com/heimdal/heimdal/issues/164#issuecomment-266231872) would go a long way towards making these build logs more readable.

So far I've debugged these issues by getting access to a machine that has one of these architectures and then reproducing the problem there. That's fairly time-consuming though.

> Also, re: `check-iprop`, a proper fix for it so that it never sleeps too little will take significant time and effort.  If this is a make-or-break issue for Debian we'll fix it and re-issue 7.1, otherwise we'd rather leave it for 7.1.1.
We can apply patches on top of the upstream release (or even disable tests altogether) - we still have a number that I haven't forwarded upstream. I am hesitant to do that without knowing what the source of the problem is though, lest we end up ignoring a critical bug that users would hit in production.

**Comment by nicowilliams** [comment]
The iprop problem requires these changes

 - `lib/kadm5/ipropdd_slave.c` needs to log after receiving _and_ applying updates (perhaps only if a verbose option is given though), preferably including the last received update vno

 - `check-iprop.in` needs to either invoke `wait_kdc` looking for a particular vno, or from a particular line number in the log (the latter is probably hard to do portably in a POSIX shell script) and _never_ sleep

Alternatively we could have an option to log a byte to a named pipe for every time `ipropd-slave` receives and applies updates, and the use `exec` to open the pipe for reading in the script.  But I'm not sure how easy it is to manage this portably in a POSIX shell.  Things like `inotify`, naturally, are out.

**Comment by nicowilliams** [comment]
I really don't know what to do about the test framework.  I find the make-driven PASS/FAIL thing horrible, but we can probably enhance it to emit debug output if `V=1`, though that too will take time and effort.

**Comment by nicowilliams** [comment]
@jelmer I'm currently running `check-iprop` from https://github.com/nicowilliams/heimdal/commit/e97706faeb88843c43d3f0d5a7454d15f47664b4 in a loop looking for failures.  So far so good.

I will push this to master.  Debian should include patch 7.1 with this.

Please review.  The main change is to add a generic shell function `wait_for_` and several functions built on top of it, `wait_for_slave_down`, `wait_for_master_down`, and `wait_for_slave`.  In particular `wait_for_slave` takes an argument specifying how the minimum change expected in the slave's  version number.

**Comment by nicowilliams** [comment]
Hmm, it may need a bit more work.  It was working when I posted that, and now it's not (a pidfile isn't getting removed).

**Comment by nicowilliams** [comment]
OK, try this commit: https://github.com/nicowilliams/heimdal/commit/0d14c82d9c0edfcc0dc33db4b2d68d24798e2696

**Comment by vdukhovni** [comment]
This still needs a bit of work I think. Failed on jessie for me. Should be fixable.

**Comment by nicowilliams** [comment]
@vdukhovni Yeah, there was one more race (I removed all the sleeps, so I made remaining races worse, which is desirable, naturally, but it took a few runs for me to hit this one).

@jelmer Try https://github.com/nicowilliams/heimdal/commit/0d14c82d9c0edfcc0dc33db4b2d68d24798e2696

**Comment by nicowilliams** [comment]
I'll leave this running.  Also, I'm running with dash(1) now in an effort to make sure this is portable.

**Comment by nicowilliams** [comment]
FYI, I pushed d237e8e to `master` to fix these races.

**Comment by jelmer** [comment]
On Thu, Dec 22, 2016 at 12:33:27PM -0800, Nico Williams wrote:
> FYI, I pushed d237e8e to `master` to fix these races.
Thanks, I'll cherry-pick these to Debian before the next upload.

**Comment by nicowilliams** [comment]
As for a test framework, there are certainly ways to improve the current test framework via the addition of suitable shell functions, but I really would prefer to be able to use ksh93 or bash, not just a POSIX shell for that, though I can write all of this in POSIX shell.  Some possible enhancements:

 - make setup-env `set -vx` if a suitable env var is set, and/or redirect `stdout` and `stderr` (so that make check can gather useful verbose failure info)
 - add a `run` function to run Heimdal executables which will:
    - keep a counter
    - run the executable under a debugger (valgrind, gdb, ...) if desired (e.g., if an env var asks for the nth run to do so)
    - prints a timestamp before and (maybe) after running the command, which would be useful for correlating to messages in the messages.log files
 - add `must_succeed`, and `must_fail` functions that call `run` and exit if the command fails or succeeds, respectively

The idea is to a) have a `make check HEIM_TESTS_VERBOSE` mode where failures produce copious useful debug information that can then be sent to us, b) improve the developer experience.

**Comment by nicowilliams** [comment]
We could also have an extra-quiet test option.  And the C-coded tests should honor the same env vars (and maybe `make` should run them via the same shell-wrapper that permits targeted debugging).

**Comment by vdukhovni** [comment]
Can this be closed now?
